### PR TITLE
BEAM-1773 Consider allowing Source#validate() to throw exception

### PR DESCRIPTION
**Pull request by tedyu**
See if this approach looks good.

### CODE REVIEW

**Comment by tedyu** [comment]
@kennknowles 
Can you take a look ?

**Comment by kennknowles** [comment]
I suppose @dhalperi and @jkff are better reviewers

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 1]
```diff
@@ -169,7 +170,7 @@ public PBegin begin() {
    * @see #apply(String, PTransform)
    */
   public <OutputT extends POutput> OutputT apply(
-      PTransform<? super PBegin, OutputT> root) {
+      PTransform<? super PBegin, OutputT> root) throws ValidationException {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: This should, at most, be added only to `Source.validate`, `PTransform.validate`, and `Pipeline.validate`. Anything calling these methods then needs to handle the exception.

Comment (tedyu) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Mind being a bit more specific on how the ValidationException should be handled ?

Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: +1, the current change is overly invasive. I would recommend to also do it in 3 separate commits (ok to do in 1 PR though) - one for Source.validate, one for PTransform.validate, one for Pipeline.validate.

Also, I think this requires more design/discussion than just adding a throws to the signature: we also need guidelines. E.g. are we going to _require_ that all errors be reported via ValidationException? I hope not - e.g. I'd like to still use Preconditions.checkArgument() for the easy stuff. If not - then which errors should be wrapped in ValidationException and which shouldn't?

Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Well, the only reason to have a `ValidationException` is that in generally a caller should treat it as alternative return code path, so "all" validation errors should proceed down that path. I agree this seems cumbersome. Your basic `main()` might just log, or a more extensive framework might put validation errors in a particular bucket separate from miscellaneous errors. But the line is blurry between validation errors and other occurrences.

@jkff you make a very good point that we probably don't need this exception. When a checked exception like `IOException` comes back from some process during `validate()` probably it should just be re-thrown as an `IllegalArgumentException` or `IllegalStateException` with a new error message explaining the context for the failure.

**Comment by aaltay** [comment]
@tedyu what is the plan for this PR?

**Pull request closed without merging**